### DISCUSSION TITLE
[@types/jest] Consider more specific types for asymmetric matchers

### DISCUSSION THREAD

**Discussion by Gaelan** [DefinitelyTyped/DefinitelyTyped#56937] [Issues with a @types package]
Currently, Jest's asymmetric matchers (like `expect.anything()`) are typed as returning any. This was done in #14609, with this code as justification:

```ts
// It's useful to create expected objects before the test call for refactoring purposes
// Assymetric matchers must return any in this case to constrain the required type
const test: Test = {
  a: expect.any(Number),
  b: expect.anything()
}
expect(callback).toHaveBeenCalledWith(test);
```

That made enough sense at the time, but it's not as type-safe as it could be (asymmetric matchers can only be used as a substitute for a value in the context of a jest assertion, but the any type would let you use them anywhere), and runs afoul of [`@eslint-typescript/no-unsafe-assignment`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-unsafe-assignment.md).

Now that we have mapped types, I think it should be possible to type this properly with something like this: 

```ts
// in jest/index.d.ts
type Matcher<T> = T | AsymmetricMatcher<T> | { [P in keyof T]: Matcher<T> };

const test: Matcher<Test> = {
  a: expect.any(Number),
  b: expect.anything()
}
expect(callback).toHaveBeenCalledWith(test);
```

This is a breaking change, though, so I'm not sure if it's worth the trouble. Thoughts?

[a comment by an automated account is not shown]

**Comment by domdomegg** [comment]
Thanks for the well thought out and detailed proposal @Gaelan.

I think this changes makes sense. Could you elaborate on what exactly are the breaking changes you expect this proposal to have? My understanding would be that people would need to change from typing it as an object and an instead a matcher for an object.

If this understanding is correct, to ease the transition we could allow both (e.g. keeping `.everything()` returning `any` for now) but promote the use of `Matcher<T>` with a mapped type (through documentation, and maybe other annotations etc.?). Then maybe with the next major update of Jest (v28 - my guess would be Jan/May 2022) also make this breaking change to the types to disable using an object (e.g. `Test`) when really it's a matcher for it (e.g. `Matcher<Test>`)?
